README.md
# mux_rtmp_flutter

A new flutter plugin project.

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.



CHANGELOG.md
## 0.0.2

- Add configure method to MuxRtmpController to allow reconnecting after disconnects

## 0.0.1

- Naive initial implementation for iOS
- Allow passing the resolution on start stream


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:mux_rtmp_flutter_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/mux_rtmp_flutter_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mux_rtmp_flutter/mux_rtmp_flutter.dart';

void main() {
  const MethodChannel channel = MethodChannel('mux_rtmp_flutter');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await MuxRtmpFlutter.platformVersion, '42');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:mux_rtmp_flutter/mux_rtmp_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    // We also handle the message potentially returning null.
    try {
      platformVersion =
          await MuxRtmpFlutter.platformVersion ?? 'Unknown platform version';
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_platformVersion\n'),
        ),
      ),
    );
  }
}


ios/Classes/MuxRtmpView.swift
import AVFoundation
import Flutter
import HaishinKit
import Logboard
import UIKit
import VideoToolbox

class MuxRtmpViewFactory: NSObject, FlutterPlatformViewFactory {
  private var messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    return MuxRtmpView(
      frame: frame,
      viewIdentifier: viewId,
      arguments: args,
      binaryMessenger: messenger
    )
  }
}

class MuxRtmpView: NSObject, FlutterPlatformView, RTMPStreamDelegate {
  private var viewId: Int64
  private var methodChannel: FlutterMethodChannel
  private var broadcastUrl: String?
  private var rtmpConnection = RTMPConnection()
  private var rtmpStream: RTMPStream!
  private var previewView: MTHKView!
  private var cameraPosition = AVCaptureDevice.Position.front

  init(
    frame: CGRect,
    viewIdentifier: Int64,
    arguments _: Any?,
    binaryMessenger: FlutterBinaryMessenger?
  ) {
    previewView = MTHKView(frame: frame)
    viewId = viewIdentifier
    methodChannel = FlutterMethodChannel(name: "mux_rtmp_view/\(viewId)", binaryMessenger: binaryMessenger!)
    super.init()
    // iOS views can be created here
    methodChannel.setMethodCallHandler {
      [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
      self?.handle(call: call, result: result)
    }
  }

  func handle(call: FlutterMethodCall, result: FlutterResult) {
    switch call.method {
    case "configure":
      if let args = call.arguments as? [String: Any] {
        configure(broadcastUrl: args["broadcastUrl"] as! String)
      }
      result(nil)
    case "startStream":
      var width = 480
      var height = 640
      if let args = call.arguments as? [String: Any] {
        if args["width"] is Int {
          width = args["width"] as! Int
        }
        if args["height"] is Int {
          height = args["height"] as! Int
        }
      }
      startStream(width: width, height: height)
      result(true)
    case "endStream":
      endStream()
      result(false)
    case "changeCamera":
      changeCamera()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func view() -> UIView {
    return previewView
  }

  func configure(broadcastUrl: String) {
    log("configure(\(broadcastUrl))")
    self.broadcastUrl = broadcastUrl
    initAVFoundation()
    initStream()
  }

  func initAVFoundation() {
    let session = AVAudioSession.sharedInstance()
    do {
      // https://stackoverflow.com/questions/51010390/avaudiosession-setcategory-swift-4-2-ios-12-play-sound-on-silent
      if #available(iOS 10.0, *) {
        try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .allowBluetooth])
      } else {
        session.perform(NSSelectorFromString("setCategory:withOptions:error:"), with: AVAudioSession.Category.playAndRecord, with: [
          AVAudioSession.CategoryOptions.allowBluetooth,
          AVAudioSession.CategoryOptions.defaultToSpeaker,
        ])
        try session.setMode(.default)
      }
      try session.setActive(true)
    } catch {
      log("initAVFoundation error: \(error)")
    }
  }

  func initStream() {
    guard let url = broadcastUrl else {
      log("initStream error: no broadcast url")
      return
    }
//    Logboard.with(HaishinKitIdentifier).level = .trace
    rtmpStream = RTMPStream(connection: rtmpConnection)

    rtmpConnection.addEventListener(.rtmpStatus, selector: #selector(rtmpStatusHandler), observer: self)
    rtmpConnection.addEventListener(.ioError, selector: #selector(rtmpErrorHandler), observer: self)

    rtmpStream.delegate = self

    rtmpStream.captureSettings = [
      .fps: 30, // FPS
      .sessionPreset: AVCaptureSession.Preset.medium, // input video width/height
      .continuousExposure: true,
      .continuousAutofocus: true,
      // .isVideoMirrored: false,
      // .continuousAutofocus: false, // use camera autofocus mode
      // .continuousExposure: false, //  use camera exposure mode
      // .preferredVideoStabilizationMode: AVCaptureVideoStabilizationMode.auto
    ]
    rtmpStream.audioSettings = [
      .muted: false, // mute audio
      .bitrate: 32 * 1000,
    ]

    // "0" means the same of input
    rtmpStream.recorderSettings = [
      AVMediaType.audio: [
        AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
        AVSampleRateKey: 0,
        AVNumberOfChannelsKey: 0,
        // AVEncoderBitRateKey: 128000,
      ],
      AVMediaType.video: [
        AVVideoCodecKey: AVVideoCodecH264,
        AVVideoHeightKey: 0,
        AVVideoWidthKey: 0,
        /*
          AVVideoCompressionPropertiesKey: [
              AVVideoMaxKeyFrameIntervalDurationKey: 2,
              AVVideoProfileLevelKey: AVVideoProfileLevelH264Baseline30,
              AVVideoAverageBitRateKey: 512000
          ]
          */
      ],
    ]

    rtmpStream.attachAudio(AVCaptureDevice.default(for: AVMediaType.audio), automaticallyConfiguresApplicationAudioSession: false) { error in
      self.log("initStream audio error \(error)")
    }
    rtmpStream.attachCamera(DeviceUtil.device(withPosition: cameraPosition)) { error in
      self.log("initStream camera error \(error)")
    }

    previewView.videoGravity = AVLayerVideoGravity.resizeAspectFill
    previewView.attachStream(rtmpStream)

    // add ViewController#view

    var bits = url.components(separatedBy: "/")
    bits.removeLast()
    rtmpConnection.connect(bits.joined(separator: "/"))
  }

  func startStream(width: Int, height: Int) {
    let uri = URL(string: broadcastUrl!)
    log("Starting stream")
    rtmpStream.videoSettings = [
      .width: width, // video output width
      .height: height, // video output height
      .bitrate: 1200 * 1000, // video output bitrate
      .profileLevel: kVTProfileLevel_H264_Baseline_AutoLevel, // H264 Profile require "import VideoToolbox"
      .maxKeyFrameIntervalDuration: 2, // key frame / sec
    ]
    rtmpStream.publish(uri?.pathComponents.last)
  }

  func endStream() {
    log("Ending stream")
    rtmpStream.close()
  }

  func changeCamera() {
    log("Changing camera")
    switch cameraPosition {
    case .back:
      cameraPosition = .front
    case .front:
      cameraPosition = .back
    default:
      cameraPosition = .back
    }

    rtmpStream.attachCamera(DeviceUtil.device(withPosition: cameraPosition)) { error in
      self.log("initStream camera error \(error)")
    }
  }

  @objc
  private func rtmpStatusHandler(_ notification: Notification) {
    log("RTMP Status Handler called.")

    let e = Event.from(notification)
    guard let data: ASObject = e.data as? ASObject, let code: String = data["code"] as? String else {
      return
    }

    log("RTMP Status: " + code)

    switch code {
    case RTMPConnection.Code.connectSuccess.rawValue:
      log("RTMP Connected")
      methodChannel.invokeMethod("rtmpStatusChange", arguments: "connected")

    case RTMPConnection.Code.connectFailed.rawValue, RTMPConnection.Code.connectClosed.rawValue:
      log("RTMP Connection was not successful.")
      methodChannel.invokeMethod("rtmpStatusChange", arguments: "notConnected")
    default:
      break
    }
  }

  // Called when there's an RTMP Error
  @objc
  private func rtmpErrorHandler(_ notification: Notification) {
    log("RTMP Error Handler called. \(notification)")
    methodChannel.invokeMethod("rtmpError", arguments: "An error occurred.")
  }

  func rtmpStreamDidClear(_: RTMPStream) {}

  // Statistics callback
  func rtmpStream(_ stream: RTMPStream, didStatics connection: RTMPConnection) {
    log("Stats: \(String(stream.currentFPS)) fps; \(String(connection.currentBytesOutPerSecond / 125)) kbps")
  }

  private var lastBwChange = 0

  // Insufficient bandwidth callback
  func rtmpStream(_ stream: RTMPStream, didPublishInsufficientBW _: RTMPConnection) {
    log("ABR: didPublishInsufficientBW")

    // If we last changed bandwidth over 10 seconds ago
    if (Int(NSDate().timeIntervalSince1970) - lastBwChange) > 5 {
      log("ABR: Will try to change bitrate")

      // Reduce bitrate by 30% every 10 seconds
      let b = Double(stream.videoSettings[.bitrate] as! UInt32) * Double(0.7)
      log("ABR: Proposed bandwidth: " + String(b))
      stream.videoSettings[.bitrate] = b
      lastBwChange = Int(NSDate().timeIntervalSince1970)

      log("Insuffient Bandwidth, changing video bandwidth to: \(String(b))")
    } else {
      log("ABR: Still giving grace time for last bandwidth change")
    }
  }

  // Today this example doesn't attempt to increase bandwidth to find a sweet spot.
  // An implementation might be to gently increase bandwidth by a few percent, but that's hard without getting into an aggressive cycle.
  func rtmpStream(_: RTMPStream, didPublishSufficientBW _: RTMPConnection) {}

  private func log(_ message: String) {
    NSLog("[MuxRtmpView] \(message)")
  }
}


android/src/main/kotlin/dev/twinsun/mux_rtmp_flutter/MuxRtmpFlutterPlugin.kt
package dev.twinsun.mux_rtmp_flutter

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** MuxRtmpFlutterPlugin */
class MuxRtmpFlutterPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "mux_rtmp_flutter")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


lib/mux_rtmp_flutter.dart
import 'dart:async';

import 'package:flutter/widgets.dart';
import 'package:flutter/services.dart';
import 'package:logging/logging.dart';

class MuxRtmpFlutter {
  static const MethodChannel _channel = MethodChannel('mux_rtmp_flutter');

  static Future<String?> get platformVersion async {
    final String? version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }
}

class MuxRtmpController {
  Logger get _log => Logger('MuxRtmpController');

  MuxRtmpController({required this.url, this.onStatusChange, this.onError});

  String url;
  void Function(String status)? onStatusChange;
  void Function(String error)? onError;

  bool _isStreaming = false;
  bool get isStreaming => _isStreaming;

  MethodChannel? _channel;

  void attachToChannel(MethodChannel channel) {
    _channel = channel;
    _channel?.setMethodCallHandler(_handle);
    configure();
  }

  Future<dynamic> _handle(MethodCall call) {
    _log.finest('${call.method}: ${call.arguments}');
    switch (call.method) {
      case 'rtmpStatusChange':
        onStatusChange?.call(call.arguments as String);
        return Future<dynamic>.value(true);
      case 'rtmpError':
        onError?.call(call.arguments as String);
        _isStreaming = false;
        return Future<dynamic>.value(true);
      default:
        return Future<dynamic>.value(null);
    }
  }

  Future<void> configure() async {
    await _channel?.invokeMethod('configure', {'broadcastUrl': url});
  }

  Future<void> startStream({double width = 480, double height = 640}) async {
    _isStreaming = await _channel?.invokeMethod('startStream', {'width': width.round(), 'height': height.round()});
  }

  Future<void> endStream() async {
    _isStreaming = await _channel?.invokeMethod('endStream');
  }

  Future<void> changeCamera() async {
    await _channel?.invokeMethod('changeCamera');
  }
}

class MuxRtmpView extends StatefulWidget {
  const MuxRtmpView({Key? key, required this.controller}) : super(key: key);

  final MuxRtmpController controller;

  @override
  State<MuxRtmpView> createState() => _MuxRtmpViewState();
}

class _MuxRtmpViewState extends State<MuxRtmpView> {
  // ignore: unused_element
  Logger get _log => Logger('MuxRtmpView');

  late final MethodChannel _channel;

  @override
  void initState() {
    super.initState();
  }

  void _handlePlatformViewCreated(int viewId) {
    _channel = MethodChannel('mux_rtmp_view/$viewId');
    widget.controller.attachToChannel(_channel);
  }

  @override
  Widget build(BuildContext context) {
    const String viewType = 'MuxRtmpView';
    final Map<String, dynamic> creationParams = <String, dynamic>{};

    return UiKitView(
      viewType: viewType,
      layoutDirection: TextDirection.ltr,
      creationParams: creationParams,
      creationParamsCodec: const StandardMessageCodec(),
      onPlatformViewCreated: _handlePlatformViewCreated,
    );
  }
}
